electronics_lib/Passives.py

```diff
@@ -314,6 +297,222 @@ def derated_capacitance(row: Dict[str, Any]) -> Tuple[float, float]:
+      max=float('inf'),
+      derate=0.04,
+      vc_pairs={100:   1e-7, 50: 4.7e-6, 25: 1e-5,             10: 2.2e-5},
+    ),
+    'Capacitor_SMD:C_1210_3225Metric': SmtCeramicCapacitorGenericPackageSpecs(
+      max=float('inf'),
+      derate=0,
+      vc_pairs={100: 4.7e-6, 50:   1e-5,           16: 2.2e-5, 10: 4.7e-5},
+    ),
+    'Capacitor_SMD:C_1812_4532Metric': SmtCeramicCapacitorGenericPackageSpecs(
+      max=float('inf'),
+      derate=0,
+      vc_pairs={100: 2.2e-6, 50:   1e-6, 25: 1e-5, },
+    ),
+  }
+
+  def select_capacitor_no_prod_table(self, capacitance: RangeVal, voltage: RangeVal,
+                                     single_nominal_capacitance: RangeVal,
+                                     part_spec: str, footprint_spec: str) -> None:
+
+    # capacitance: user-specified capacitance
+    # single nominal capacitance: no single cap with requested capacitance, must generate multiple parallel caps
+    # nominal capacitance: selected part's capacitance
+
+    def valid_package_min_nominal_capacitance(package: int, capacitance: RangeVal, voltage: RangeVal) -> Tuple[float, float]:
+      """
+      Calculates if the package is a valid choice for the given capacitance and voltage.
+      If the package is a valid choice, the function returns the minimum nominal capacitance
+      for the specs. If the package is not valid, the function returns False.
+      """
+      avg_cap = (capacitance[0] + capacitance[1]) / 2
+      avg_volt = (voltage[0] + voltage[1]) / 2
+
+      # Calculate derated capacitance for the package
+      if (avg_volt <= 3.6) or (avg_cap <= 1e-6):
+        min_nominal_capacitance = capacitance
+      else:
+        voltco = self.PACKAGE_SPECS[package].derate
+        min_nominal_capacitance = (
+          capacitance[0] / (1 - voltco * (voltage[1] - 3.6)),
+          capacitance[1] / (1 - voltco * (voltage[0] - 3.6))
+        )
+
+      # Check if package is still valid for the derated capacitance
+      if self.PACKAGE_SPECS[package].max >= avg_cap:
+        for v, c in self.PACKAGE_SPECS[package].vc_pairs.items():
+          if v >= voltage[1] and c >= min_nominal_capacitance[1]:
+            return min_nominal_capacitance
+      return False
+
+    def select_package(capacitance: RangeVal, voltage: RangeVal) -> Tuple[float, float]:
+
+      if footprint_spec == "":
+        for package in sorted(self.PACKAGE_SPECS.keys()):
+          package_derated_capacitance = valid_package_min_nominal_capacitance(package, capacitance, voltage)
+          if package_derated_capacitance:
+            return (package, package_derated_capacitance)
+        return ("", (0, 0))
+      else:
+        package_derated_capacitance = valid_package_min_nominal_capacitance(footprint_spec, capacitance, voltage)
+        if package_derated_capacitance:
+          return (footprint_spec, package_derated_capacitance)
+      return (footprint_spec, (0, 0))
```

Comment:
Can you unify this block with the case where `footprint_spec == ""`? One way I can think of is to create a package_options, which is `[footprint_spec]` (if `footprint_spec` is not empty) or `self.PACKAGE_SPECS.keys()` (if `footprint_spec` is empty)